kms/src/main/java/io/kroxylicious/kms/provider/kroxylicious/inmemory/InMemoryKms.java

```diff
@@ -0,0 +1,222 @@
+    private final KeyGenerator aes;
+    private final int numIvBytes;
+    private final int numAuthBits;
+    private final SecureRandom secureRandom;
+    private final Map<String, UUID> aliases;
+
+    InMemoryKms(int numIvBytes, int numAuthBits,
+                Map<UUID, SecretKey> keys,
+                Map<String, UUID> aliases) {
+        this.keys = keys;
+        this.aliases = aliases;
+        this.secureRandom = new SecureRandom();
+        this.numIvBytes = numIvBytes;
+        this.numAuthBits = numAuthBits;
+        try {
+            this.aes = KeyGenerator.getInstance(KEY_ALGO);
+        }
+        catch (NoSuchAlgorithmException e) {
+            // This should be impossible, because JCA guarantees that AES is available
+            throw new KmsException(e);
+        }
+    }
+
+    /**
+     * Generates a KEK
+     * @return The id of the KEK
+     */
+    public UUID generateKey() {
+        var key = aes.generateKey();
+        var ref = UUID.randomUUID();
+        keys.put(ref, key);
+        return ref;
+    }
+
+    public void createAlias(UUID keyId, String alias) {
+        lookupKey(keyId); // check the key exists in this KMS
+        aliases.put(alias, keyId);
+    }
+
+    public void deleteAlias(String alias) {
+        var was = aliases.remove(alias);
+        if (was == null) {
+            throw new UnknownAliasException(alias);
+        }
+    }
+
+    @NonNull
+    @Override
+    public CompletableFuture<InMemoryEdek> generateDek(@NonNull UUID kekRef) {
+        try {
+            return CompletableFuture.completedFuture(wrap(kekRef, this.aes::generateKey));
+        }
+        catch (KmsException e) {
+            return CompletableFuture.failedFuture(e);
+        }
+    }
+
+    private InMemoryEdek wrap(UUID kekRef, Supplier<SecretKey> generator) {
+        SecretKey kek = lookupKey(kekRef);
+        Cipher aesCipher = aesGcm();
+        GCMParameterSpec spec = aesGcmSpec(kek, aesCipher);
+        var dek = generator.get();
+        byte[] edek;
+        try {
+            edek = aesCipher.wrap(dek);
+        }
+        catch (IllegalBlockSizeException | InvalidKeyException e) {
+            throw new KmsException(e);
+        }
+        return new InMemoryEdek(spec.getTLen(), spec.getIV(), edek);
+    }
+
+    @NonNull
+    @Override
+    public CompletableFuture<DekPair<InMemoryEdek>> generateDekPair(@NonNull UUID kekRef) {
+        try {
+            var dek = this.aes.generateKey();
+            var edek = wrap(kekRef, () -> dek);
+            return CompletableFuture.completedFuture(new DekPair<>(edek, dek));
+        }
+        catch (KmsException e) {
+            return CompletableFuture.failedFuture(e);
+        }
+    }
+
+    private GCMParameterSpec aesGcmSpec(SecretKey kek, Cipher aes) {
+        byte[] iv = new byte[numIvBytes];
+        secureRandom.nextBytes(iv);
+        var spec = new GCMParameterSpec(numAuthBits, iv);
+        try {
+            aes.init(Cipher.WRAP_MODE, kek, spec);
```

Comment:
I'd have the caller of aesGcmSpec do the initialisation leaving this method just to create the GCMParameterSpec